Support Gemma Models

cmathw opened the pull request:
# Description

Added support for google/gemma model series: [gemma-2b](https://huggingface.co/google/gemma-2b), [gemma-2b-it](https://huggingface.co/google/gemma-2b-it), [gemma-7b](https://huggingface.co/google/gemma-7b) and [gemma-7b-it](https://huggingface.co/google/gemma-7b-it). The technical report associated with these models can be found [here](https://storage.googleapis.com/deepmind-media/gemma/gemma-report.pdf).

## Notes
This model's architecture is very similar to Mistral with a few key differences: 
- Does not use sliding window attention
- d_head > d_model // n_head for 7b model
- RMSNorm is slightly different wherein 1 is added to the RMSNorm weights before multiplying by input.
- Embeddings are scaled by sqrt(d_model).

For the 2b model, logits are now within 5e-4, all resid_pre caches are within 1e-4 and next token loss within 5e-6. A demo  showing this agreement can be found [here](https://colab.research.google.com/drive/1i_PB509R9Dy8xmqIkyZ9XkXMvX3wSpm0?usp=sharing). I haven't tested on the 7b model due to memory constraints.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

neelnanda-io commented:
The 1 + weights thing in RMS Norm is very weird! Can we resolve this by just editing the weights during the loading_from_pretrained.py step?

cmathw commented:
> The 1 + weights thing in RMS Norm is very weird! Can we resolve this by just editing the weights and manually adding 1 during the loading_from_pretrained.py step? That way there's no need to actually change the code.

Yep I think that is ok, that is what I have done currently. There is still a difference between cache and logits I am investigating though.

ArthurConmy commented:
![image](https://github.com/neelnanda-io/TransformerLens/assets/35957051/d7fb6882-5974-4810-8b2d-b58f07f2a952)

May be due to post-attention normalization

cmathw commented:
> ![image](https://private-user-images.githubusercontent.com/35957051/306795497-d7fb6882-5974-4810-8b2d-b58f07f2a952.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NTU2MzMsIm5iZiI6MTcwODU1NTMzMywicGF0aCI6Ii8zNTk1NzA1MS8zMDY3OTU0OTctZDdmYjY4ODItNTk3NC00ODEwLThiMmQtYjU4ZjA3ZjJhOTUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDIyNDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3MjRkNTE2NmI1ZDg4ZDRlZjgzYWIwNTA2YmY2YWVkODU1NGI5MDJhZjQxZDkwYzQ0NDhkODc2NTVlY2YwZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.O8BMTwF_6SqUdT_1Ij-1KoYROivMZPZvwSgfVXHKr0I)
> 
> May be due to post-attention normalization
> 
> EDIT: assuming https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma/modeling_gemma.py#L636 is correct, this screenshot is just calling the pre-MLP layer norm "post_attention_layer_norm", i.e. your issue is probably not due to normalization weirdness

Yeah it seems pretty standard in the transformers repo, I'll keep digging!

neelnanda-io commented:
I'm very confused by that line about pre and post LN, it looks to just be pre LN in both HuggingFace and our internal implementation. If you get good logits from the HuggingFace version (eg next token loss <4 on natural language) then I'd assume the paper is incorrect.

cmathw commented:
A couple of quick questions around this PR:

- I've needed to update the to_str_tokens method to add a batch dimension as the Gemma Tokeniser expects a 2D tensor when calling `batch_decode.` I expect this to also be needed with other tokenizers (ie. LLaMA) but I haven't tested this. Is this something that anyone has run into before when trying to use `batch_decode ` or `to_str_tokens` in the past?

- The gemma models require logging into hugging face via `huggingface-cli` or a providing token, how do we typically deal with this when adding a model to TL? At the moment, this is PR is failing to build docs due to not being able to access the HF config for this model.

cmathw commented:
> I'm very confused by that line about pre and post LN, it looks to just be pre LN in both HuggingFace and our internal implementation. If you get good logits from the HuggingFace version (eg next token loss <4 on natural language) then I'd assume the paper is incorrect.

Yeah I think the paper is just misleading, a similar implementation is on the Google repo now: https://github.com/google/gemma_pytorch/blob/5a160cb8755edc334c3bb751267d3e012d3446bb/gemma/model.py#L300

ojh31 commented:
Any reason not to include the instruct models here? https://huggingface.co/google/gemma-2b-it https://huggingface.co/google/gemma-7b-it

neelnanda-io commented:
Curious how things are going here Chris, and if you need a hand?

On Mon, 11 Mar 2024, 1:31 pm Oskar John Hollinsworth, <
***@***.***> wrote:

> Any reason not to include the instruct models here?
> https://huggingface.co/google/gemma-2b-it
> https://huggingface.co/google/gemma-7b-it
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/neelnanda-io/TransformerLens/pull/511#issuecomment-1988450442>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASRPNKOX5YED6BUFZM4RLJDYXWW4LAVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBYGQ2TANBUGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

cmathw commented:
> Curious how things are going here Chris, and if you need a hand?
> […](#)
> On Mon, 11 Mar 2024, 1:31 pm Oskar John Hollinsworth, < ***@***.***> wrote: Any reason not to include the instruct models here? https://huggingface.co/google/gemma-2b-it https://huggingface.co/google/gemma-7b-it — Reply to this email directly, view it on GitHub <[#511 (comment)](https://github.com/neelnanda-io/TransformerLens/pull/511#issuecomment-1988450442)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ASRPNKOX5YED6BUFZM4RLJDYXWW4LAVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBYGQ2TANBUGI> . You are receiving this because you commented.Message ID: ***@***.***>

@neelnanda-io This PR is pretty close to finished and ready for review, I've checked tolerances for the 2b and 2b instruct tuned models and these are [fine](https://colab.research.google.com/drive/1i_PB509R9Dy8xmqIkyZ9XkXMvX3wSpm0?usp=sharing). The main thing that is left is that I am not sure how we typically deal with 'gated' models that need huggingface-cli login to access, the `build docs` CI is currently failing due to this. How do we typically deal with this?

neelnanda-io commented:
I'd just hard-code the config so the docs can load, and people will still
need to log in to access the weights

On Mon, 11 Mar 2024 at 17:35, cmathw ***@***.***> wrote:

> Curious how things are going here Chris, and if you need a hand?
> … <#m_8946512441480891079_>
> On Mon, 11 Mar 2024, 1:31 pm Oskar John Hollinsworth, < *@*.*> wrote: Any
> reason not to include the instruct models here?
> https://huggingface.co/google/gemma-2b-it
> <https://huggingface.co/google/gemma-2b-it>
> https://huggingface.co/google/gemma-7b-it
> <https://huggingface.co/google/gemma-7b-it> — Reply to this email directly,
> view it on GitHub <#511 (comment)
> <https://github.com/neelnanda-io/TransformerLens/pull/511#issuecomment-1988450442>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/ASRPNKOX5YED6BUFZM4RLJDYXWW4LAVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBYGQ2TANBUGI
> <https://github.com/notifications/unsubscribe-auth/ASRPNKOX5YED6BUFZM4RLJDYXWW4LAVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBYGQ2TANBUGI>
> . You are receiving this because you commented.Message ID: @.*>
>
> @neelnanda-io <https://github.com/neelnanda-io> This PR is pretty close
> to finished and ready for review, I've checked tolerances for the 2b and 2b
> instruct tuned models and these are fine
> <https://colab.research.google.com/drive/1i_PB509R9Dy8xmqIkyZ9XkXMvX3wSpm0?usp=sharing>.
> The main thing that is left is that I am not sure how we typically deal
> with 'gated' models that need huggingface-cli login to access, the build
> docs CI is currently failing due to this. How do we typically deal with
> this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/neelnanda-io/TransformerLens/pull/511#issuecomment-1989040204>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASRPNKLWVFDSQYVGQPIXHWDYXXTO5AVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBZGA2DAMRQGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cmathw commented:
This should be ready to merge in after a review!

neelnanda-io commented:
Thanks! This overall looks good to me. One nitpick: I find the "multiply
embedding scale by sqrt(d_model)" fairly inelegant, thoughts on instead
multiplying W_E by this when doing the weight conversion, so we don't need
the config option explicitly?

My main concern is that this'll make the logits have higher error,
especially in low precision. It'll also break the fact that W_E and W_U are
tied, but that's not a big deal (and is normally broken by folding or
centering W_U anyway)

On Thu, 14 Mar 2024 at 16:39, cmathw ***@***.***> wrote:

> This should be ready to merge in after a review!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/neelnanda-io/TransformerLens/pull/511#issuecomment-1997874427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASRPNKPDQQ7OD7UENTAO7ZTYYHHDJAVCNFSM6AAAAABDT3VBJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJXHA3TINBSG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cmathw commented:
> Thanks! This overall looks good to me. One nitpick: I find the "multiply embedding scale by sqrt(d_model)" fairly inelegant, thoughts on instead multiplying W_E by this when doing the weight conversion, so we don't need the config option explicitly? My main concern is that this'll make the logits have higher error, especially in low precision. It'll also break the fact that W_E and W_U are tied, but that's not a big deal (and is normally broken by folding or centering W_U anyway)

Yep that makes sense, I have updated the `convert_gemma_weights` to reflect this and removed this change from `components.py`.

neelnanda-io commented on transformer_lens/HookedTransformerConfig.py:
```diff
@@ -151,6 +151,8 @@ class HookedTransformerConfig:
             Only for models that use Grouped Query Attention.
         post_embedding_ln (bool): Whether to apply layer normalization after embedding the tokens. Defaults
             to False.
+        post_embedding_scale (bool): Whether to scale the post-embedding tokens by multiplying by sqrt(d_model).
```
Can you also remove it from the config?

cmathw replied:
Apologies for missing this, fixed now!

neelnanda-io commented:
Amazing, thanks so much for adding this!
